Dockerfile

```diff
@@ -1,29 +1,115 @@
-FROM ruby:2.7.2
-
-# 1. Configuring apt to properly install yarn
-# 2. Installing node, postgres-client, yarn
-# 3. Clearing apt-cached files to tighten the image size
-RUN curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | APT_KEY_DONT_WARN_ON_DANGEROUS_USAGE=1 apt-key add - \
-  && echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list \
-  && apt-get update -qq \
-  && apt-get install -y build-essential \
-  nodejs \
-  yarn \
-  && rm -rf /var/lib/apt/lists/*
+# Generating a generic Rails image.
+# Just for caching.
+FROM ruby:2.7-alpine AS rails
 
-WORKDIR /app
+# `apk` is like `apt`, but for Alpine Linux.
+RUN apk add --update --no-cache --quiet \
+    build-base \
+    nodejs-current \
+    postgresql-dev \
+    sqlite-dev \
+    tzdata \
+    yarn
+
+# Create a generic Gemfile with only Rails.
+RUN echo -e "source 'https://rubygems.org'\ngem 'rails'" > /Gemfile
+# Install Rails, quietly, in 4 threads.
+RUN bundle install --quiet -j4
+
+# Uses the image above as the base for a builder image.
+# This will **exclude** development and testing gems.
+FROM rails as rails-builder
+
+# Set the value of `pwd` for all following operations.
+WORKDIR /app/
+
+# Add the Gemfile and Gemfile.lock from this application.
+COPY Gemfile Gemfile.lock /app/
+
+# Installs gems, quietly, in 4 threads, and retry if broken.
+RUN bundle install --quiet -j4 --retry 3 \
+    # Remove unneeded files (cached *.gem, *.o, *.c)
+    && rm -rf /usr/local/bundle/cache/*.gem \
+    && find /usr/local/bundle/gems/ -name "*.c" -delete \
+    && find /usr/local/bundle/gems/ -name "*.o" -delete
+
+# Expose port 5000 from the final image.
+# This is important if you deploy to Gitlab, otherwise it does
+# not matter which port we use here.
+EXPOSE 5000
+
+# When starting this container, run the rails server
+# binding to all IP's, and port 5000.
+CMD bundle exec rails server -b 0.0.0.0 -p 5000
```

Comment:
Why are we using 5000 and not 3000?

Reply:
Port 5000 is important for Gitlab, but does not matter for most other environments, thus it felt like a sane default.

In addition, the `docker-compose.yml` maps this port to port 3000 for local development to match with Rails standard.

Reply:
So, we use 3000 for local development and 5000 for production??

Reply:
How does this CMD work if we haven't copied in the app dir and Gemfile?

Reply:
Because this image is used in `docker-compose.yml`, the `CMD` is meant to be executed after mounting the current working directory as a volume.  HuntersKeepers adds some additional commands (such as `db:migrate`), which override this base command.  However, this command is a guideline for those commands.

Both the `EXPOSES` and `CMD` can be removed until the `final` image if they are confusing.

Reply:
I would greatly appreciate moving those because it would be much less confusing for me.

I now understand how the node_modules and app are loaded by the volume in docker-compose.yml.